elastic/kibana
# [Security solution] Generative AI Connector

## GAI Connector: Supports OpenAI and Azure OpenAI

New Generative AI connector that does not fit in any of the current connector categories. It has its own unique type, named `GeneralFeature` for now as we attempt to better define the function of this type of connector. 

Allows user to define a generative AI connector with provider type OpenAI or Azure OpenAI.
Current only has the most basic executor action: send request to the chat completion endpoint and receive back the response.

Example OpenAI payload:
```
{
  "model": "gpt-3.5-turbo",
  "messages": [  {  "role": "user", "content": "Hello world" } ]
}
```

Example Azure OpenAI payload:
```
{
  "messages": [  {  "role": "user", "content": "Hello world" } ]
}
```

### Define connector
<img width="1200" alt="Screenshot 2023-05-12 at 2 55 29 PM" src="https://github.com/elastic/kibana/assets/6935300/2af35d90-5aa7-4786-abcc-d3a220ebd395">
<img width="1200" alt="Screenshot 2023-05-12 at 2 55 18 PM" src="https://github.com/elastic/kibana/assets/6935300/42e97a71-d02c-4013-96e8-b6c035e3fb84">
<img width="1200" alt="Screenshot 2023-05-12 at 2 54 54 PM" src="https://github.com/elastic/kibana/assets/6935300/b201b46c-f9e5-4277-a356-6dcc7032d645">



### Test connector: OpenAI
<img width="1200" alt="Screenshot 2023-05-12 at 9 08 40 AM" src="https://github.com/elastic/kibana/assets/6935300/b9dcfab4-0b27-47d0-89eb-717cbfa8ac61">

### Test connector: Azure OpenAI
<img width="1200" alt="Screenshot 2023-05-12 at 9 17 49 AM" src="https://github.com/elastic/kibana/assets/6935300/f519578e-9c6f-4fa9-ae7e-c54a9d5d8c4b">

## Testing
Credentials for OpenAI and Azure OpenAI: https://p.elstc.co/paste/DmPeZJUW#dZBlr35dZvYbxpnu0vylJXJd6OJqZ1FiGOgdDalmymV
1. Create GAI connectors with OpenAI & Azure OpenAI credentials
2. Test each connector by sending example payload
3. Confirm results in Network tab. Should look like:
<img width="877" alt="Screenshot 2023-05-12 at 9 37 53 AM" src="https://github.com/elastic/kibana/assets/6935300/9b9ecd86-0d56-42ab-ad97-024527f6251b">
